Incorrect suggestion for incompatible match arm types

### Code

```Rust
use std::ops::ControlFlow;

fn main() {
    let mut copy: Vec<U> = three_arg_diff(1, X {}, X {});

    match x {
        Err(r) => ControlFlow::Break(r),
        2_000_000..=3_999_999 => { println!("A")}
    }
}
```

### Current output

```Shell
error[E0308]: `match` arms have incompatible types
 --> .\2\test.rs:8:36
  |
6 | /     match x {
7 | |         Err(r) => ControlFlow::Break(r),
  | |                   --------------------- this is found to be of type `ControlFlow<_, _>`
8 | |         2_000_000..=3_999_999 => { println!("A")}
  | |                                    ^^^^^^^^^^^^^ expected `ControlFlow<_, _>`, found `()`
9 | |     }
  | |_____- `match` arms have incompatible types
  |
  = note:   expected enum `ControlFlow<_, _>`
          found unit type `()`
  = note: this error originates in the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
help: try wrapping the expression in a variant of `ControlFlow`
 --> C:\Users\jjl98\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\macros.rs:142
:23
  |
14~     ($($arg:tt)*) => {std::ops::ControlFlow::Continue({
14|         $crate::io::_print($crate::format_args_nl!($($arg)*));
14~     })};
 --> C:\Users\jjl98\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\macros.rs:142
:23
  |
14~     ($($arg:tt)*) => {std::ops::ControlFlow::Break({
14|         $crate::io::_print($crate::format_args_nl!($($arg)*));
14~     })};
  |
```

### Desired output

```Shell
error[E0308]: `match` arms have incompatible types
  --> .\2\test.rs:9:36
   |
6  | /     match x {
7  | |     //~^ ERROR cannot move out of `x` as enum variant `Some` which is behind a shared reference
8  | |         Err(r) => ControlFlow::Break(r),
   | |                   --------------------- this is found to be of type `ControlFlow<_, _>`
9  | |         2_000_000..=3_999_999 => { println!("A")}
   | |                                    ^^^^^^^^^^^^^ expected `ControlFlow<_, _>`, found `()`
10 | |     }
   | |_____- `match` arms have incompatible types
   |
   = note:   expected enum `ControlFlow<_, _>`
           found unit type `()`
   = note: this error originates in the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Rationale and extra context

The suggestion here is incorrect — users should not be prompted to modify code in the standard library. In fact, it would be better not to show any suggestion at all, as was the case in version 1.86. The "Desired Output" shown above is exactly what rustc 1.86.0 (05f9846f8 2025-03-31) produces.

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.89.0-nightly (1677d46cb 2025-06-10)
binary: rustc
commit-hash: 1677d46cb128cc8f285dbd32b0dc4d7a46437050
commit-date: 2025-06-10
host: x86_64-pc-windows-msvc
release: 1.89.0-nightly
LLVM version: 20.1.5
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -2707,12 +2707,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     Some(CtorKind::Const) => unreachable!("unit variants don't have fields"),
                 };
 
-                // Suggest constructor as deep into the block tree as possible.
-                // This fixes https://github.com/rust-lang/rust/issues/101065,
-                // and also just helps make the most minimal suggestions.
+                // Suggest constructor as deep into the block tree as possible,
+                // but don't cross macro contexts. This fixes #101065 while
+                // keeping suggestions out of macro definitions (#142359).
                 let mut expr = expr;
                 while let hir::ExprKind::Block(block, _) = &expr.kind
                     && let Some(expr_) = &block.expr
+                    && expr_.span.eq_ctxt(expr.span)
                 {
                     expr = expr_
                 }

--- a/tests/ui/typeck/suggestions/auxiliary/suggest-compatible-variants-macro-issue-142359.rs
+++ b/tests/ui/typeck/suggestions/auxiliary/suggest-compatible-variants-macro-issue-142359.rs
@@ -1,0 +1,6 @@
+#[macro_export]
+macro_rules! unit {
+    () => {{
+        ()
+    }};
+}

--- a/tests/ui/typeck/suggestions/suggest-compatible-variants-macro-issue-142359.rs
+++ b/tests/ui/typeck/suggestions/suggest-compatible-variants-macro-issue-142359.rs
@@ -1,0 +1,16 @@
+// Make sure we don't suggest compatible variants cross macro context. (issue #142359)
+//@ aux-crate:ext=suggest-compatible-variants-macro-issue-142359.rs
+
+extern crate ext;
+
+use std::ops::ControlFlow;
+
+fn main() {
+    let x: Result<i32, i32> = Err(1);
+
+    let _ = match x {
+        Err(r) => ControlFlow::Break(r),
+        Ok(r) => { ext::unit!() } //~ ERROR `match` arms have incompatible types [E0308]
+
+    };
+}

--- a/tests/ui/typeck/suggestions/suggest-compatible-variants-macro-issue-142359.stderr
+++ b/tests/ui/typeck/suggestions/suggest-compatible-variants-macro-issue-142359.stderr
@@ -1,0 +1,25 @@
+error[E0308]: `match` arms have incompatible types
+  --> $DIR/suggest-compatible-variants-macro-issue-142359.rs:13:19
+   |
+LL |       let _ = match x {
+   |  _____________-
+LL | |         Err(r) => ControlFlow::Break(r),
+   | |                   --------------------- this is found to be of type `ControlFlow<i32, _>`
+LL | |         Ok(r) => {ext::unit!()}
+   | |                   ^^^^^^^^^^^^ expected `ControlFlow<i32, _>`, found `()`
+LL | |
+LL | |     };
+   | |_____- `match` arms have incompatible types
+   |
+   = note:   expected enum `ControlFlow<i32, _>`
+           found unit type `()`
+   = note: this error originates in the macro `ext::unit` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: try wrapping the expression in `std::ops::ControlFlow::Continue`
+  --> $DIR/auxiliary/suggest-compatible-variants-macro-issue-142359.rs:4:9
+   |
+LL |         std::ops::ControlFlow::Continue(())
+   |         ++++++++++++++++++++++++++++++++  +
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/typeck/suggestions/suggest-compatible-variants-macro-issue-142359.stderr
+++ b/tests/ui/typeck/suggestions/suggest-compatible-variants-macro-issue-142359.stderr
@@ -1,12 +1,12 @@
 error[E0308]: `match` arms have incompatible types
-  --> $DIR/suggest-compatible-variants-macro-issue-142359.rs:13:19
+  --> $DIR/suggest-compatible-variants-macro-issue-142359.rs:13:20
    |
 LL |       let _ = match x {
    |  _____________-
 LL | |         Err(r) => ControlFlow::Break(r),
    | |                   --------------------- this is found to be of type `ControlFlow<i32, _>`
-LL | |         Ok(r) => {ext::unit!()}
-   | |                   ^^^^^^^^^^^^ expected `ControlFlow<i32, _>`, found `()`
+LL | |         Ok(r) => { ext::unit!() }
+   | |                    ^^^^^^^^^^^^ expected `ControlFlow<i32, _>`, found `()`
 LL | |
 LL | |     };
    | |_____- `match` arms have incompatible types
@@ -15,10 +15,9 @@ LL | |     };
            found unit type `()`
    = note: this error originates in the macro `ext::unit` (in Nightly builds, run with -Z macro-backtrace for more info)
 help: try wrapping the expression in `std::ops::ControlFlow::Continue`
-  --> $DIR/auxiliary/suggest-compatible-variants-macro-issue-142359.rs:4:9
    |
-LL |         std::ops::ControlFlow::Continue(())
-   |         ++++++++++++++++++++++++++++++++  +
+LL |         Ok(r) => std::ops::ControlFlow::Continue({ ext::unit!() })
+   |                  ++++++++++++++++++++++++++++++++                +
 
 error: aborting due to 1 previous error
 

